### ISSUE TITLE
[BUG] (v0.0.0): Prelude_No._1_BWV_846_in_C_Major.mxl

### ISSUE THREAD

**Issue by rvilarl** [stringsync/vexml#207]
**Expected behavior**
Render correctly.
![image](https://github.com/stringsync/vexml/assets/22865285/d5e4bbd0-3919-43eb-a549-e986421da830)

**Actual behavior**
Exception and wrong rendering.

**Screenshot**
![image](https://github.com/stringsync/vexml/assets/22865285/a9ce6152-73e6-4cb2-8d4a-2156e81734a3)

**Hints**
The more you provide, the faster we can fix it.

- [x] Attach the MusicXML file used to produce the bug.
- [x] Attach the error messages.
- [x] Include any `vexml` code references where the bug may be.

**Comment by rvilarl** [comment]
[Prelude_No._1_BWV_846_in_C_Major.zip](https://github.com/stringsync/vexml/files/14072898/Prelude_No._1_BWV_846_in_C_Major.zip)

**Comment by rvilarl** [comment]
@jaredjj3 @ronyeh I am moving to another city. Once I settle I will be more active. :)

**Comment by jaredjj3** [comment]
@rvilarl thanks for the bug report and good luck with your move!

**Comment by jaredjj3** [comment]
Looking at this now.

**Comment by jaredjj3** [comment]
To troubleshoot, I colored the stave notes being used for the start and end notes.

```ts
  render(): PedalRendering {
    const vfStaveNotes = this.getVfStaveNotes();
    const vfPedalMarkingType = this.getVfPedalMarkingType();
    const vfPedalMarking = new vexflow.PedalMarking(vfStaveNotes).setType(vfPedalMarkingType);

    vfStaveNotes[0].setStyle({ fillStyle: 'green' }); // Change the color to green
    vfStaveNotes[1].setStyle({ fillStyle: 'red' }); // Change the color to red

    return {
      type: 'pedal',
      vexflow: {
        pedalMarking: vfPedalMarking,
      },
    };
  }
  ```
  
  
<img width="700" alt="image" src="https://github.com/stringsync/vexml/assets/19232300/7ccb3293-ecc7-41cf-8dc0-19e7582c205d">

It looks like the second note of the second pedal marking is not being associated correctly. Even if the same note was being used as a start and end, I expect it to show as the end color.

When I implemented `<direction>` support, I always attached it to a note, since that's what `vexflow` needs. However, I think I might need to _conditionally_ create `TextNote` to avoid issues like this.

I'm going to spend some time figuring out a better strategy for `<direction>` elements, especially before I try to do lilypond31a (Directions).

**Comment by jaredjj3** [comment]
I thought I could make `vexflow.TextNote` objects that have no ticks, but that seems to be a special case for `vexflow.BarNote` objects ([code](https://github.com/0xfe/vexflow/blob/8ddc8fa1a6d304a879e73830919fa17f3a9bdef4/src/tables.ts#L29)). I'm skeptical that even if there was a supported way to render `vexflow.TextNote` as such, it would probably interfere with formatting. I'm going to study the problem a bit more.

@rvilarl, @infojunkie, and @sschmidTU, do you have any advice for handling `<direction>` elements? They're inherently not attached to a specific `<note>`, but `vexflow` typically requires a `vexflow.Tickable` to render them. It seems like I might have to handle each direction on a case-by-case basis — there doesn't seem to be a universal strategy I can apply to them like `<technical>` or `<articulations>`.

**Comment by jaredjj3** [comment]
For now, this is what I'm planning to do when iterating through a `<measure>`:

1. When handling a `<note>`, attach all consecutive `<direction>` elements after it.
2. When handling a `<direction>` without a preceding `<note>` in scope, attach it to the next `<note>` within that measure.
3. If there is no `<note>` within the `<measure>`, create a `vexflow.GhostNote` and attach the `<directions>` to it. If this doesn't work, hack an empty `TextNote` with duration `'256'` instead.

**Comment by infojunkie** [comment]
Looks like a complex specification indeed. And of course, it is complex because of all the cases of annotations that exist in real-word scores.

My reading of the [`<direction>` specification](https://www.w3.org/2021/06/musicxml40/musicxml-reference/elements/direction/) tells me the following factors would affect the positioning algorithm:

- The optional sub-element `<offset>` indicates where the direction will appear relative to the current musical location (in unit of score divisions).
- The optional attribute `directive` changes the default-x position of a direction.
- By default, a series of `<direction-type>` elements and a series of child elements of a `<direction-type>` within a single `<direction>` element follow one another in sequence visually.
- For applications where a specific direction is indeed attached to a specific note, the `<direction>` element can be associated with the first `<note>` element that **follows it in score order** that is not in a different voice.
- Child elements of `<direction-type>` include attributes `default-x`, `default-y`, `relative-x`, `relative-y` that affect their individual positioning.

At this moment, I don't fully understand which direction types need to be associated with notes, versus which aren't. Maybe it's just a function of where they occur in the bar, but that would need clarification. Maybe you can ask a question in the [musicxml discussion forum](https://github.com/w3c/musicxml/discussions) although the new maintainer of the format has not been very responsive. You may have better luck at [StackExchange Music](https://music.stackexchange.com/).

Happy to continue digging if you feel this is useful to you.

**Comment by rvilarl** [comment]
@jaredjj3 I also do not see like @infojunkie why we have to associate directions to notes. Probably the abstraction can be improved in VexFlow, for example, Coda & Segno are StaveModifiers but pedal is not.

**Comment by jaredjj3** [comment]
@infojunkie, thank you, this is an excellent summarization (even better than ChatGPT).

>- For applications where a specific direction is indeed attached to a specific note, the <direction> element can be associated with the first <note> element that follows it in score order that is not in a different voice.

Ah, of course! This is where I went wrong — the current `vexml` algorithm associates a `<direction>` with the first succeeding `<note>` regardless of `<voice>`. 

Here's the snippet I was dealing with (the full score is in https://github.com/stringsync/vexml/issues/207#issuecomment-1913191681):

```
<note>
  <voice>5</voice>
</note>
<direction placement="below">
  <direction-type>
    <pedal type="stop" line="yes"/>
    </direction-type>
  <staff>2</staff>
  </direction>
<backup>
  <duration>16</duration>
</backup>
<note>
  <voice>6</voice>
</note>
```

The `<direction>` actually needs to be associated with the previous `<note>`, because the next `<note>` of the same `<voice>` appears in the next measure. The [MusicXML `<measure>` spec](https://www.w3.org/2021/06/musicxml40/musicxml-reference/elements/measure-partwise/) also permits a `<direction>` to be specified without any `<note>`.

**Comment by jaredjj3** [comment]
> @jaredjj3 I also do not see like @infojunkie why we have to associate directions to notes. Probably the abstraction can be improved in VexFlow, for example, Coda & Segno are StaveModifiers but pedal is not.

I think this is a shortcoming of the MusicXML spec. It's very confusing to have directives that _maybe_ associate `<direction>` and `<note>` elements. Hopefully it'll be improved in [mnx](https://w3c.github.io/mnx/docs/).

Ultimately, yes I agree if `vexflow` had a way to add `StaveModifier` objects **inline** within a `Voice`, it would make this a lot easier. I'm going to see how hard the implementation is before submitting an FR to `vexflow`. Thanks for calling that out!

---

The more I work with `vexflow`, I think this project would've been significantly easier if I was able to attach non-tickables to the stave in a matrix-like fashion. I'm not asking for this to be put in `vexflow` since it's really invasive, but I would like to share my thoughts.

```ts
const stave1 = Stave({
  label: Text('part1', { offset: Vector.fromOrigin(-5, 0) }),
  entries: [
    // A StaveEntry is like a vertical stack of elements.
    StaveEntry([Barline({ label: Text('1') })]),
    StaveEntry([KeySignature('C')]),
    StaveEntry([Clef('G')]),
    StaveEntry([TimeSignature('4/4')]),
    StaveEntry([Note('C', { duration: 'h' })]),
    StaveEntry([Chord([Note('C'), Note('D')])]),
  ]
});
const stave2 = Stave(); // empty for now
const system = System({ staves: [stave1, stave2] });
system.format().draw();
```

or maybe you have helpful instance methods:

```ts
const stave = Stave();

// This would merge into the internal structures of the stave.
stave.insert(StaveEntry([Note('C', duration: 'q')]), { at: Fraction(3, 4) });
```

This would be a different goal than `EasyScore`. This API is not optimizing for user writability — it's optimizing for programability.

I'm not sure how adaptable this is to `vexml` and if the benefits are worth the effort. I'll just keep this in the back of my mind next time I get blocked.

**Comment by infojunkie** [comment]
> @infojunkie, thank you, this is an excellent summarization (even better than ChatGPT).

You have a low opinion of humans and a high opinion of technology 😱

**Comment by jaredjj3** [comment]
I'm waiting on https://github.com/0xfe/vexflow/issues/1611 before I resume this.

**Comment by jaredjj3** [comment]
Fixed for all viewport sizes  in v0.1.2.

![Image](https://github.com/user-attachments/assets/a6344b9d-5a25-45d1-9e73-359caa5b87de)
